src/util/checkpt/fd_checkpt.h

```diff
@@ -0,0 +1,540 @@
+   optimizations.  Should be at most FD_CHECKPT_PRIVATE_CHUNK_USZ_MAX.
+   Set to 0 to disable checkpt small gather optimization. */
+
+#define FD_CHECKPT_PRIVATE_GBUF_THRESH FD_CHECKPT_PRIVATE_CHUNK_USZ_MAX
+
+/* FD_CHECKPT_PRIVATE_GBUF_SZ is the size of the checkpt gather buffer.
+   Must be at least 2*THRESH + 65536 - 3 (the 64KiB is from LZ4). */
+
+#define FD_CHECKPT_PRIVATE_GBUF_SZ (2UL*FD_CHECKPT_PRIVATE_GBUF_THRESH + 65536UL)
+
+FD_STATIC_ASSERT( FD_CHECKPT_PRIVATE_GBUF_THRESH <= FD_CHECKPT_PRIVATE_CHUNK_USZ_MAX,             adjust_gbuf_limits );
+FD_STATIC_ASSERT( FD_CHECKPT_PRIVATE_GBUF_SZ     >= 2UL*FD_CHECKPT_PRIVATE_GBUF_THRESH + 65533UL, adjust_gbuf_limits );
+
+/* FD_RESTORE_PRIVATE_SBUF_{THRESH,SZ} similarly give the configuration
+   for small buffer restore scatter optimizations.  These currently need
+   to match the gather configuration. */
+
+#define FD_RESTORE_PRIVATE_SBUF_THRESH FD_CHECKPT_PRIVATE_GBUF_THRESH
+#define FD_RESTORE_PRIVATE_SBUF_SZ     FD_CHECKPT_PRIVATE_GBUF_SZ
+
+/* fd_checkpt_t internals */
+
+struct fd_checkpt_private_wbuf { /* very similar to fd_io_buffered_ostream */
+  uchar * mem;  /* Buffer of compressed bytes not yet written to fd, byte indexed [0,wbuf_sz) */
+  ulong   sz;   /* Buffer size in bytes, >=FD_CHECKPT_WBUF_MIN */
+  ulong   used; /* Buffer bytes [0,wbuf_used) are not yet written to fd,
+                   buffer bytes [wbuf_used,wbuf_sz) are free */
+};
+
+typedef struct fd_checkpt_private_wbuf fd_checkpt_private_wbuf_t;
+
+struct fd_checkpt_private_mmio {
+  uchar * mem; /* Checkpoint memory region */
+  ulong   sz;  /* Checkpoint memory region byte size */
+};
+
+typedef struct fd_checkpt_private_mmio fd_checkpt_private_mmio_t;
+
+struct fd_checkpt_private {
+  int    fd;          /* (stream) File descriptor for the checkpt (>=0), (mmio) -1 */
+  int    frame_style; /* FD_CHECKPT_FRAME_STYLE_* (>0), 0: not in frame (valid), -1: not in frame (failed) */
+  void * lz4;         /* Handle of the underlying compressor */
+  ulong  gbuf_cursor; /* Cursor for small buffer gather optimizations, in [0,FD_CHECKPT_PRIVATE_GBUF_SZ] */
+  ulong  off;         /* Offset of the next byte to write (relative to the checkpoint first byte), in [0,mmio_sz) in mmio mode */
+  union {
+    fd_checkpt_private_wbuf_t wbuf; /* used in streaming mode */
+    fd_checkpt_private_mmio_t mmio; /* used in mmio mode */
+  };
+  uchar gbuf[ FD_RESTORE_PRIVATE_SBUF_SZ ]; /* gather optimization buffer */
+};
+
+/* fd_restore_t internals */
+
+struct fd_restore_private_rbuf { /* very similar to fd_io_buffered_istream */
+  uchar * mem;  /* Buffer of compressed bytes read from fd, byte indexed [0,rbuf_sz) */
+  ulong  sz;    /* Buffer size in bytes, >=FD_RESTORE_RBUF_MIN */
+  ulong  lo;    /* Buffer bytes [0,rbuf_lo) have been read and restored */
+  ulong  ready; /* Number of compressed bytes that haven't been processed, 0<=rbuf_lo<=(rbuf_lo+rbuf_ready)<=rbuf_sz */
```

Comment:
alignment

Reply:
None required (or really beneficial in this situation)